# Factorial Function in Python

#### Objective: To practice creating Jupyter Notebooks, writing and running Python code, documenting with Markdown, and uploading the notebook to GitHub.



This notebook demonstrates:
- How to define a simple Python function
- Using recursion to calculate factorial
- Documenting with Markdown cells

In [1]:

def factorial(n):
    if n == 0 or n == 1:
        return 1
    else:
        return n * factorial(n - 1)

In [3]:

print(factorial(5))

120


Base case: If the number is 0 or 1, return 1.
(Because 0! = 1 and 1! = 1 by definition.)

Recursive case: If n is bigger than 1, the function calls itself with n - 1.
So factorial(5) → 5 × factorial(4) → 5 × 4 × factorial(3) … until it reaches factorial(1).

## How the Factorial Function Works

The factorial of a number `n` is the product of all positive integers up to `n`.  
For example:  
**5! = 5 × 4 × 3 × 2 × 1 = 120**

We used a **recursive approach**:
- If `n` is 0 or 1, return 1.  
- Otherwise, return `n × factorial(n - 1)`.  

This continues until the base case (1 or 0) is reached.

### Why Recursion Works Here

Each time the function calls itself, the problem becomes smaller (n gets reduced by 1).
Eventually, it reaches the base case (1), and then the results are multiplied back together.